raster-foundry/raster-foundry
# Buffer requests to api/scenes

## Overview

Configure buffer sizes for `POST` requests to `/api/scenes` using the Nginx `client_body_buffer_size` parameter to make sure that ~99% of requests won't get buffered to disk.

### Checklist


## Demo

100kb Sentinel request without `client_body_buffer_size`:

```
2019/02/21 18:09:47 [warn] 6#6: *8 a client request body is buffered to a temporary file /var/cache/nginx/client_temp/0000000001, client: 10.0.2.2, server: app.staging.rasterfoundry.com, request: "POST /api/scenes/ HTTP/1.1", host: "localhost:9100"
10.0.2.2 - - "POST /api/scenes/ HTTP/1.1" 201 100786 - "-" "RFLambda 0.1.0-08eed316" 0.798 0.800
```

That same request with the buffer size set to `256k`:

```
10.0.2.2 - - "POST /api/scenes/ HTTP/1.1" 201 100786 - "-" "RFLambda 0.1.0-08eed316" 0.175 0.172
```

## Notes

For more background on the investigation that led to my choices in this PR, see: https://github.com/azavea/raster-foundry-platform/issues/631#issuecomment-465758035

## Testing Instructions

You can test the buffering behavior of the Lambda function locally:

- Save a test event to `app-lambda/test-event.json` that mimics a 100k Sentinel 2 SNS topic:

```json
{
  "Records": [
    {
      "EventSource": "aws:sns",
      "EventVersion": "1.0",
      "EventSubscriptionArn": "arn:aws:sns:eu-west-1:214830741341:NewSentinel2Product:7f07ce2f-fe89-4ace-be5e-3cb6118a246d",
      "Sns": {
        "Type": "Notification",
        "MessageId": "2283514d-9c1a-5cf7-b387-05f5b4c9f03a",
        "TopicArn": "arn:aws:sns:eu-west-1:214830741341:NewSentinel2Product",
        "Subject": "S2A_MSIL1C_20190115T145721_N0207_R039_T19MBP_20190115T164135",
        "Message": "{\n  \"name\" : \"S2A_MSIL1C_20190115T145721_N0207_R039_T19MBP_20190115T164135\",\n  \"id\" : \"eec370a1-4da8-4ed7-9313-125098191c94\",\n  \"path\" : \"products/2019/1/15/S2A_MSIL1C_20190115T145721_N0207_R039_T19MBP_20190115T164135\",\n  \"timestamp\" : \"2019-01-15T14:57:21.024Z\",\n  \"datatakeIdentifier\" : \"GS2A_20190115T145721_018626_N02.07\",\n  \"sciHubIngestion\" : \"2019-01-15T18:12:52.990Z\",\n  \"s3Ingestion\" : \"2019-01-15T18:28:55.782Z\",\n  \"tiles\" : [ {\n    \"path\" : \"tiles/19/M/BP/2019/1/15/0\",\n    \"timestamp\" : \"2019-01-15T15:04:39.013Z\",\n    \"utmZone\" : 19,\n    \"latitudeBand\" : \"M\",\n    \"gridSquare\" : \"BP\",\n    \"datastrip\" : {\n      \"id\" : \"S2A_OPER_MSI_L1C_DS_SGS__20190115T164135_S20190115T145721_N02.07\",\n      \"path\" : \"products/2019/1/15/S2A_MSIL1C_20190115T145721_N0207_R039_T19MBP_20190115T164135/datastrip/0\"\n    }\n  } ],\n  \"datastrips\" : [ {\n    \"id\" : \"S2A_OPER_MSI_L1C_DS_SGS__20190115T164135_S20190115T145721_N02.07\",\n    \"path\" : \"products/2019/1/15/S2A_MSIL1C_20190115T145721_N0207_R039_T19MBP_20190115T164135/datastrip/0\"\n  } ]\n}",
        "Timestamp": "2019-01-15T18:29:39.373Z",
        "SignatureVersion": "1",
        "Signature": "LZgGjZWAiNeByMlGgR80Vu4jTLbGZpqgQ0ctddBP1RFGQdjR3zlaYq56hhCFC6l2wTi2JIpcqqnSDqbq4wRgwchteakQ7oFbhjInRByY+4geHWthk+tIfSqm36WsbF5S1OcsJ2FhzqAh2PfF334ybs966B3omZPVZnYRfPsE3Sxt03p5XQSWCORnxUhKw6oJQPH0RD6kj/kIvOfzps/hSoSVLuhv/OhjR1iEDDKC8rjOC7apVKOTgVCWLLSu0M8A0T0WSK57d4akaEntCxjqumCv8/N0scan4hVXMP1ZWaMR4IVyP3eZ0OkNx8v4NVosEIp+iab7nmugikPBsPPMvA==",
        "SigningCertUrl": "https://sns.eu-west-1.amazonaws.com/SimpleNotificationService-ac565b8b1a6c5d002d285f9598aa1d9b.pem",
        "UnsubscribeUrl": "https://sns.eu-west-1.amazonaws.com/?Action=Unsubscribe&SubscriptionArn=arn:aws:sns:eu-west-1:214830741341:NewSentinel2Product:7f07ce2f-fe89-4ace-be5e-3cb6118a246d",
        "MessageAttributes": {}
      }
    }
  ]
}
```

- Start a development server with `./scripts/server`
  - I was getting a lot of extra output, so I ran all services except for `api-server` and `nginx-api` in the background
- Change to the `app-lambda` directory with `cd app-lambda`
- Export some environment variables that the Lambda function needs:
  - `AWS_PROFILE=raster-foundry`
  - `RF_API_HOST=localhost:9100`
  - `RF_API_TOKEN` (or `RF_REFRESH_TOKEN`, the token that the app frontend gives you)
  - `GIT_COMMIT=$(git rev-parse --short HEAD)`, if you want to test the interpolation of the commit SHA; otherwise, the version string in the request's User-Agent should be set to `0.1.0-dev`
- Create a Python virtualenv: `python -m venv venv; source venv/bin/activate`
- Install the Lambda function with `pip install .`
- Run the following code in a Python interpreter:

```python
import json
from rflambda.sentinel2.handle import handle
js = json.load(open('test-event.json', 'r'))
handle(js, {})
```

- If you'd like, you can also test the behavior without increased buffers by commenting out the new directive in `nginx/etc/nginx/conf.d/api.conf`:

```diff
--- a/nginx/etc/nginx/conf.d/api.conf
+++ b/nginx/etc/nginx/conf.d/api.conf
@@ -30,7 +30,7 @@ server {
     location /api/scenes {
         include /etc/nginx/includes/proxy-settings.conf;
         proxy_buffers 24 4k;
-        client_body_buffer_size 256k;
+        # client_body_buffer_size 256k;

         proxy_pass http://api-server-upstream;
     }
```

Closes https://github.com/azavea/raster-foundry-platform/issues/631 and closes #4639.
